### ISSUE TITLE
cmd/go: follow symlinks referenced by //go:embed

### ISSUE THREAD

**Issue by fmeum** [golang/go#59924]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.20.3 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

N/A, applies to any GOOS/GOARCH

### What did you do?

In a Bazel build action running in a symlink-based sandbox, try to build the Go standard library or any third-party package using `//go:embed`.

### What did you expect to see?

The build passes just as it would if the input files weren't symlinks. Tools only play nice with Bazel if they follow symlinks, as whether or not an input file to a Bazel build action is a symlink depends on the particular strategy with which Bazel chooses to execute the action (e.g. remotely or locally, in a sandbox or not, ...).

`//go:embed` not allowing symlinks to be embedded makes sense and I am not asking for that to be changed. However, it would be very helpful if it could be made to *follow* symlinks, either by default or optionally, guarded by a flag or environment variable.

### What did you see instead?

```
external/go_sdk/src/crypto/elliptic/p256_asm.go:24:12: pattern p256_asm_table.bin: cannot embed irregular file p256_asm_table.bin
```

This happens because of the use of `Lstat` in https://github.com/golang/go/blob/3c46d8f5119475adbadb6141acd463b910c1747b/src/cmd/go/internal/load/pkg.go#L2139, which makes symlinks hit the branch at https://github.com/golang/go/blob/3c46d8f5119475adbadb6141acd463b910c1747b/src/cmd/go/internal/load/pkg.go#L2170

This issue requires rules_go to [copy around the stdlib](https://github.com/bazelbuild/rules_go/commit/db9458406295b6f569fc1fa5142483636fb86b45) and [breaks build actions that operate on generated code](https://github.com/bazelbuild/rules_go/issues/3178).

**Comment by fmeum** [comment]
CC @thanm

**Comment by thanm** [comment]
@golang/tools-team

**Comment by fmeum** [comment]
@thanm This continues to be a problem. If there is anything I can do, such as submit a PR to modify this behavior based on an environment variable, please let me know.

**Comment by codenoid** [comment]
assign me on this task please, I face the same issue

**Comment by thequailman** [comment]
I worked around this by excluding the symlink, e.g.:

```
//go:embed mydir/[^excludeme]*
```

And doing another embed for the underlying symlink path.

**Comment by bcmills** [comment]
@fmeum, I'm no longer maintaining `cmd/go`. (I left Google last March.)

**Comment by thanm** [comment]
Hi @fmeum as it turns out I am also no longer with Google as of last July. Sorry I can't help --

**Comment by fmeum** [comment]
@golang/tools-team Friendly ping

**Comment by matloob** [comment]
@fmeum To clarify: the files being embedded are themselves symlinks right? In general we've tried to avoid following symlinks within modules because doing so adds complexity we'd like to try to avoid.

**Comment by fmeum** [comment]
Yes, Bazel's sandbox represents all files as symlinks to absolute paths. I can see the security concerns with allowing arbitrary files to be embedded during a `go build`, but being able to explicitly allow symlinks to be followed would help with this situation.

**Comment by matloob** [comment]
I discussed this offline with @rsc @samthanawalla. We don't want to allow symlinks in the general case, especially symlinks to other directories, but we can make this case work. We'll modify the check to allow symlinks to regular files.

**Comment by nightlyone** [comment]
Do the targets of those symlinks still have to be located below or at the same level in  the file system hierarchy as the directory of the file containing the //go:embed pragma?

So can a symlink in a dependency I download embed a symlink to /etc/password ?

**Comment by rsc** [comment]
GODEBUG=embedfollowsymlinks=1 ?

**Comment by cespare** [comment]
We have a large Go module (a monorepo) with a bunch of Go webservers and we like to use `//go:embed` to bundle up assets (JS, CSS) into the binaries. In some cases we have some shared assets (say, vendored JS files) that would be nice to share between different servers within that module. Symlinking is one obvious way to accomplish this. So if there were security limits on the link targets, it would be nice to restrict symlinking outside of the *module*, not outside of the directory of the file with the `//go:embed` comment.

(Although for the Bazel use case, I'm guessing it does need to symlink outside of the module?)

**Comment by matloob** [comment]
Yes, for the Bazel use case we'd have to symlink outside the module. They usually don't need the go command because they invoke the compiler and linker themselves. But, since we stopped shipping .a files for the standard library, they need to use the go command to produce the .a files for the standard library. As I understand it, in the symlink sandbox all the files the action encounters are symlinked, so they'd necessarily have to be outside of GOROOT.
